Simplify diffing

elizabethengelman opened the pull request:
Update:
Wanted to make sure to note that when testing this out with VDB, though the core interface hasn't changed, some of the data types have changed (i.e. AccountDiff, StorageDiff, etc have few fields now) so its important to update the geth version in the vdb project as well. 

-------

This is still a work in progress, but I'd love some early feedback on the approach. Trying out getting the account and storage diffs from stateObjects right before the changes are committed via a call to the method `Commit`. 
- Part of the rationale for this is that in the `Commit` step the trie nodes are potentially flushed to disk and/or pruned if it's a non-archive node. When the nodes are pruned or flushed, we no longer have access to them in-memory, so we'd run into errors in the statediff building because it couldn't look up necessary node. Previously to combat this, we had tweaked the pruning algorithm. But based on feedback in http://github.com/ethereum/go-ethereum/pull/19916 it's preferable to not have to update the pruning since it's already pretty complex.
- Another potential upside to getting state diffs this way is that we're reducing the need to traverse the state and storage tries a second time for the diffing... here we are getting any state that is deemed "dirty" or changed when the transaction are being processed anyway.


Still left to do:
- [x] add tests for getting modified accounts 
- [x] remove statediff builder?
- [x] allow filtering by watched addresses
- [ ] is there a better way to handle the subscription than registering a service?
- [ ] compare these results with the results from Parity and the old geth patch method

elizabethengelman committed:
Temporarily remove IntermediateNodes from statdiffing service

elizabethengelman committed:
Temporarily remove streaming a block alongside state diffs

elizabethengelman committed:
Temporarily remove streaming paths and proofs with state diffs

elizabethengelman committed:
Remove changes to pruning algo

This means that the current implementation will only work with an
archive node.

elizabethengelman committed:
WIP - sending block info before commit

was hoping that this would avoid the issue of trie nodes being pruned
before we had the chance to process their diffs (with a full node) but I
am still seeing those errors with this.

elizabethengelman committed:
WIP - return modified state and storage data from Commit

elizabethengelman committed:
WIP - sending StateChangeEvents to the statediffing service

which includes a block and modified accounts

elizabethengelman committed:
Remove PrecommitedChainEvent subscription in favor of a StateChangeEvent sub

elizabethengelman committed:
WIP - use StateChangeEvents feed instead of ChainEvents feed

elizabethengelman committed:
Move Block attribute from StateChanges to StateChangeEvent

elizabethengelman committed:
Clean up statediff blockchain interface

elizabethengelman committed:
Cleanup and apply gofmt changes

elizabethengelman committed:
Update statediff service_test to include account and storage data

rmulhol commented on core/blockchain.go:
```diff
@@ -1298,10 +1289,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
-	if err != nil {
-		return NonStatTy, err
+	root, modifiedAccounts, commitErr := state.Commit(bc.chainConfig.IsEIP158(block.Number()))
+	log.Info("Sending StateChangeEvent to the feed", "block number", block.Number(), "count", len(modifiedAccounts.ModifiedAccounts))
```
Totally non-blocking but I wonder if `Trace` or `Debug` might be more appropriate here, assuming this happens pretty frequently

elizabethengelman replied:
Yep, totally! Nice catch.

rmulhol commented on core/state/statedb.go:
```diff
@@ -724,33 +724,55 @@ func (s *StateDB) clearJournalAndRefund() {
 			}
 		}
+
+		modifiedAccounts.ModifiedAccounts[addr] = modifiedAccount
```
Just to make sure I understand - is it correct that a deleted account will result in the payload mapping the address to an empty `ModifiedAccount`?

elizabethengelman replied:
I think that's correct that `Commit` would return a mapping from an address to an empty `ModifiedAccount`. I recently made a change to not return Payload with empty `ModifiedAccounts`, so that we're not spamming the subscribers with empty payloads. But I'd be happy to think that through a bit more.

rmulhol commented on statediff/service.go:
```diff
@@ -147,32 +125,49 @@ func (sds *Service) Loop(chainEventCh chan core.ChainEvent) {
+			storageDiffs = append(storageDiffs, diff)
+		}
+
+		address := addr
```
What's the purpose of this assignment?

elizabethengelman replied:
Great question. When I didn't assign `addr` to a new variable, I was seeing some state weirdness, where it seemed like as we iterated through the `modifiedAccounts`, the address would be overwritten in the returned `accountDiffs` collection. So they'd all have the same address for some reason. I haven't fully figured it out yet, but I'll probably spend some more time digging in to see if there's a better solution.

rmulhol commented on statediff/service.go:
```diff
@@ -147,32 +125,49 @@ func (sds *Service) Loop(chainEventCh chan core.ChainEvent) {
+		BlockHash:       stateChangeEvent.Block.Hash(),
+		UpdatedAccounts: accountDiffs,
+		encoded:         nil,
+		err:             nil,
```
Are these keys used elsewhere? Wondering if we can remove them from the struct

elizabethengelman replied:
I'm not entirely sure how/where we are using them yet, but I agree if they're no longer used, it would make sense to remove them.

rmulhol reviewed (approved):
Nice! This looks really awesome - does it mean that we no longer need to delay pruning to perform additional lookups against the db? Exciting stuff!

elizabethengelman committed:
Make log statement for each new StateChangeEvent Debug instead of Info

elizabethengelman committed:
Remove statediff builder

Instead of iterating through the existing state trie, this approach gets
dirty state objects right after the transactions have been processed

elizabethengelman committed:
Use dirtyStateObject.originStorage instead of dirtyStateObject.dirtyStorage

elizabethengelman committed:
StorageDiff Value should be an RLP encoded object

elizabethengelman committed:
Add statedb test for stateChanges emitted from Commit

elizabethengelman committed:
Remove unused test helpers

elizabethengelman committed:
Remove Encode interface from Payload struct

the Payloads aren't being rlp encoded before sending to the subscription, so the encode
interface isn't currently necessary

elizabethengelman commented on statediff/types.go:
```diff
@@ -39,27 +38,6 @@ type Payload struct {
-	}
-}
-
-// Length to implement Encoder interface for Payload
```
We're not currently encoding the Payloads, so I removed the Encoder interface because it wasn't being used. I'm wondering if rlp encoding the Payloads is required for this first pass. 🤔

elizabethengelman committed:
Remove unused statediff type fields and helper methods

elizabethengelman committed:
Update log level to Debug for sending diff payloads to subscription

elizabethengelman committed:
Refactor processAccountDiffs - breaks it up into smaller methods

elizabethengelman committed:
Refactor StateChanges type from a struct to a map

this removes an uncessary layer in the data structure

elizabethengelman committed:
Filter statediffs by watched addresses if they're configured

elizabethengelman committed:
Refactor statediff service test

elizabethengelman committed:
Don't send a payload if the AccountDiff is empty

elizabethengelman committed:
Update account in statedb_test

only have one account have a storage change

elizabethengelman committed:
Wire up watchedaddresses flag to the statediff service

elizabethengelman committed:
Get checksum of watchedconfigured address before comparing to current address

elizabethengelman committed:
Switch stateChangeEventCh logging from Info to Debug

elizabethengelman commented:
>does it mean that we no longer need to delay pruning to perform additional lookups against the db?

Yep! We're grabbing the changes from [state.Commit ](https://github.com/makerdao/go-ethereum/blob/d49601cff775a1dfb593f9544d9f52a272b20930/core/blockchain.go#L1292) before the [garbage collection](https://github.com/makerdao/go-ethereum/blob/d49601cff775a1dfb593f9544d9f52a272b20930/core/blockchain.go#L1308) is happening.

elizabethengelman commented:
closing this PR in favor of #11
